### PR TITLE
short circuit in if/case_when function

### PR DESCRIPTION
**Pull request by zengyan1**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #14989

## What this PR does / why we need it:
short circuit in if function

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Added a `selectList` parameter to the `Eval` function calls across multiple files.
- Implemented short-circuit evaluation for the `IFF` function in `evalExpression.go`.
- Updated various `Eval` function implementations to accept the `selectList` parameter.
- Updated tests to include `selectList` parameter in `Eval` function calls.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><details><summary>14 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>mysql_cmd_executor.go</strong><dd><code>Add `selectList` parameter to `Eval` function calls.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pkg/frontend/mysql_cmd_executor.go

- Added a `selectList` parameter to the `Eval` function calls.

</details>
    

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/16521/files#diff-af2611d5fc89704398fe09d09644efa41fec8931b395eda292f2f474f1216275">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>join.go</strong><dd><code>Add `selectList` parameter to `Eval` function calls.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pkg/sql/colexec/anti/join.go

- Added a `selectList` parameter to the `Eval` function calls.

</details>
    

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/16521/files#diff-1ed87fe1fad1db0822265563079fd991a1f006f2a8392ce3551d1cbba51391da">+3/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>evalExpression.go</strong><dd><code>Add `selectList` parameter and implement short-circuit evaluation.</code></dd></summary>
<hr>
      
pkg/sql/colexec/evalExpression.go

<li>Added <code>selectList</code> parameter to <code>Eval</code> function.<br> <li> Implemented short-circuit evaluation for <code>IFF</code> function.<br> <li> Updated various <code>Eval</code> function implementations to accept <code>selectList</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/16521/files#diff-096074b4e06b58b1b2144cc4b4d91fe38b5c4ced8bfd165e7cfc66bd31aa341d">+82/-17</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>external.go</strong><dd><code>Add `selectList` parameter to `Eval` function calls.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pkg/sql/colexec/external/external.go

- Added a `selectList` parameter to the `Eval` function calls.

</details>
    

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/16521/files#diff-81231a121d9293d51f08e00a14fd08a5d7f50631fc9cea82a67940ab0f2f78b3">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>filter.go</strong><dd><code>Add `selectList` parameter to `Eval` function calls.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pkg/sql/colexec/filter/filter.go

- Added a `selectList` parameter to the `Eval` function calls.

</details>
    

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/16521/files#diff-bc93911f832b77a7a27b34ce9d74f1325641a3fd0923e0450ab5622e0ecb7d94">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>group.go</strong><dd><code>Add `selectList` parameter to `Eval` function calls.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pkg/sql/colexec/group/group.go

- Added a `selectList` parameter to the `Eval` function calls.

</details>
    

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/16521/files#diff-7dbdbce3ed6c3f03c5a3bd5419acf07e900f03ff08ebcc118e26ee5fa4611bd8">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>build.go</strong><dd><code>Add `selectList` parameter to `Eval` function calls.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pkg/sql/colexec/hashbuild/build.go

- Added a `selectList` parameter to the `Eval` function calls.

</details>
    

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/16521/files#diff-0c2753ffdc4ccc951ac44650817da1d8ffe37631363a440b8b55732bb0e2b839">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>join.go</strong><dd><code>Add `selectList` parameter to `Eval` function calls.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pkg/sql/colexec/join/join.go

- Added a `selectList` parameter to the `Eval` function calls.

</details>
    

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/16521/files#diff-2a43ca762c59d0ece91404da915aa519c35db896df1ec228aab36a10e7dbc1dd">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>join.go</strong><dd><code>Add `selectList` parameter to `Eval` function calls.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pkg/sql/colexec/left/join.go

- Added a `selectList` parameter to the `Eval` function calls.

</details>
    

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/16521/files#diff-e504c243155528e43dfd3b5e88ce6641dd41fbe3db3d5b6a0ee7d607bca91b3d">+3/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>limit.go</strong><dd><code>Add `selectList` parameter to `Eval` function calls.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pkg/sql/colexec/limit/limit.go

- Added a `selectList` parameter to the `Eval` function calls.

</details>
    

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/16521/files#diff-a11c08586a47d366d79a6d3640723a37d19f5cc14507d73282e66c854551ce27">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>join.go</strong><dd><code>Add `selectList` parameter to `Eval` function calls.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pkg/sql/colexec/loopanti/join.go

- Added a `selectList` parameter to the `Eval` function calls.

</details>
    

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/16521/files#diff-726bedc6ba50fdcc528d7bc44e42646aaee4f81af2e357b32c0d57929efb49c0">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>join.go</strong><dd><code>Add `selectList` parameter to `Eval` function calls.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pkg/sql/colexec/loopjoin/join.go

- Added a `selectList` parameter to the `Eval` function calls.

</details>
    

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/16521/files#diff-e41ade36a55ecf6b0b2215fb6b319a079529707f6f65a64a897ba7979cbb74b0">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>join.go</strong><dd><code>Add `selectList` parameter to `Eval` function calls.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pkg/sql/colexec/loopleft/join.go

- Added a `selectList` parameter to the `Eval` function calls.

</details>
    

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/16521/files#diff-16bf2fa1511b121867abcf2fdc9d0c93fea49617a965d0e9778c71cd974b60c5">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>join.go</strong><dd><code>Add `selectList` parameter to `Eval` function calls.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pkg/sql/colexec/loopmark/join.go

- Added a `selectList` parameter to the `Eval` function calls.

</details>
    

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/16521/files#diff-10d77a4cb110eb09f811731ddd2c5793f55fbe112742114acbfe4af401e2dc0d">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests
</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>evalExpression_test.go</strong><dd><code>Update tests for `Eval` function with `selectList` parameter.</code></dd></summary>
<hr>
      
pkg/sql/colexec/evalExpression_test.go

<li>Updated tests to include <code>selectList</code> parameter in <code>Eval</code> function calls.<br>

</details>
    

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/16521/files#diff-f98b8d63c73a672a92da3a404fce145662b7e2b268962b505cd3c6f686c03acc">+6/-6</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Review by XuPeng-SH — APPROVED** [review]

**Review by qingxinhome — APPROVED** [review]

**Review by badboynt1 — APPROVED** [review]

File: pkg/sql/colexec/evalExpression.go [hunk 1]
```diff
@@ -477,19 +480,89 @@ func (expr *FunctionExpressionExecutor) Init(
...
+			selectList = make([]bool, batches[0].RowCount())
+			for i := range selectList {
+				selectList[i] = true
+			}
+		}
+		expr.parameterResults[0], err = expr.parameterExecutor[0].Eval(proc, batches, selectList)
+		if err != nil {
+			return nil, err
+		}
+		selectList1 := make([]bool, batches[0].RowCount())
```
Comment (m-schen) [inline pkg/sql/colexec/evalExpression.go:496]: 这个可以存在 FunctionExpressionExecutor里复用起来。这样多次计算期间不需要重新开辟。

File: pkg/sql/colexec/evalExpression.go [hunk 2]
```diff
@@ -70,7 +70,7 @@ type ExpressionExecutor interface {
 	// Eval will return the result vector of expression.
 	// the result memory is reused, so it should not be modified or saved.
 	// If it needs, it should be copied by vector.Dup().
-	Eval(proc *process.Process, batches []*batch.Batch) (*vector.Vector, error)
+	Eval(proc *process.Process, batches []*batch.Batch, selectList []bool) (*vector.Vector, error)
```
Comment (m-schen) [inline pkg/sql/colexec/evalExpression.go:73]: 下面那个EvalWithoutResultReusing也顺便改掉吧，两个接口保持一致。

File: pkg/sql/colexec/evalExpression.go [hunk 3]
```diff
@@ -477,19 +480,89 @@ func (expr *FunctionExpressionExecutor) Init(
 	return err
 }
 
-func (expr *FunctionExpressionExecutor) Eval(proc *process.Process, batches []*batch.Batch) (*vector.Vector, error) {
+func (expr *FunctionExpressionExecutor) Eval(proc *process.Process, batches []*batch.Batch, selectList []bool) (*vector.Vector, error) {
 	var err error
-	for i := range expr.parameterExecutor {
-		expr.parameterResults[i], err = expr.parameterExecutor[i].Eval(proc, batches)
+	if expr.fid == function.IFF {
```
Comment (m-schen) [inline pkg/sql/colexec/evalExpression.go:485]: 这块代码最好单独拉出去作为一个独立的方法。
if isIFF := fid == function.IFF; isIFF {
    if err := f(); err != nil { return err}
}
提高代码的可读性。

File: pkg/sql/colexec/evalExpression.go [hunk 4]
```diff
@@ -477,28 +482,116 @@ func (expr *FunctionExpressionExecutor) Init(
 	return err
 }
 
-func (expr *FunctionExpressionExecutor) Eval(proc *process.Process, batches []*batch.Batch) (*vector.Vector, error) {
+func (expr *FunctionExpressionExecutor) EvalIff(proc *process.Process, batches []*batch.Batch, selectList []bool) (err error) {
+	if selectList == nil {
+		selectList = make([]bool, batches[0].RowCount())
```
Comment (m-schen) [inline pkg/sql/colexec/evalExpression.go:487]: 这样比没复用性能更差了，还多开辟了一个selectList.

File: pkg/sql/plan/function/func_builtin.go [hunk 5]
```diff
@@ -226,7 +226,7 @@ func builtInMoShowVisibleBinEnum(parameters []*vector.Vector, result vector.Func
 	for i := uint64(0); i < uint64(length); i++ {
 		v1, null1 := p1.GetStrValue(i)
 		enumStr, null2 := enumVal.GetStrValue(i)
-		if null1 || null2 || len(v1) == 0 || len(enumStr) == 0 {
+		if null1 || null2 || len(v1) == 0 || len(enumStr) == 0 || result.GetResultVector().GetNulls().Contains(i) {
```
Comment (m-schen) [inline pkg/sql/plan/function/func_builtin.go:229]: 我觉得可以给 result 这个结构体新增一个方法，
func IgnoreCompute(idx) bool, 
或者 func AlreadySet(idx) bool之类的方法，
它的作用就是判断是否为其nsp是否contain().
这样代码更容易理解，也可以免去这边的3层函数调用。

同时可以做进一步优化，就是在生成result的MustFunctionResult方法里直接让他变成两种结构体，
一种是每次都去检查nsp是否contains的，
一种是始终返回false的，这样可以避免每次调用都去判断nsp != nil这件事。

其他函数以及baseTemplate同理。

File: pkg/sql/plan/function/baseTemplate.go [hunk 6]
```diff
@@ -434,30 +434,18 @@ func decimalArith[T templateDec](parameters []*vector.Vector, result vector.Func
...
-					}
-					r, err := arithFn(v1, v2, scale1, scale2)
-					if err != nil {
-						return err
-					}
-					rss[i] = r
+			nulls.Or(rsVec.GetNulls(), parameters[1].GetNulls(), rsVec.GetNulls())
+			rowCount := uint64(length)
+			for i := uint64(0); i < rowCount; i++ {
+				if rsVec.GetNulls().Contains(i) {
```
Comment (m-schen) [inline pkg/sql/plan/function/baseTemplate.go:440]: 与普通函数相同，这里要尽量减少关于nsp的判断，
可以通过提前筛选出nsp == nil的情况或者类似其他评论说的给result提供IgnoreCompute()等方法。

可以多思考一下，或者直接改变整个模板的实现。

我感觉把函数的传参变成
parameters, result, length, ignore 似乎也是可以的。其中ignore是一个结构体的指针（直接从function expression里存的selectList1, selectList2组成），对外提供 shouldIgnore方法，
这样模板里可以通过判断这个是否为空来省很多操作，普通函数也会好改一些。
并且可以省去执行框架里每次把selectList变成nsp的步骤，当然这个步骤的成本被转移到了函数实现里的append null，但是目前的代码改动其实也会做这步冗余操作。

作为对函数开发者提供的模板，可以多做一些优化。
（这里可以考虑把baseTemplate变成由MakeFile生成，可以减轻一些修改压力和误改，不过这个是另一件事情了）。

**Review by m-schen — CHANGES_REQUESTED** [review]
我觉得可以再多思考一下。现在主要问题是：

1. 执行框架中 selectList 的复用问题。
2. 函数具体实现里，突然对result的vector判断nsp在该pr中可以明白目的，但后续会难以理解。
3. 函数具体实现（包括模板）可以做一些性能上的优化。

然后最好再跑跑各种性能测试。

File: pkg/sql/plan/function/func_builtin.go (hunk 5 above)
Comment (m-schen) [reply pkg/sql/plan/function/func_builtin.go:229]: 只是提议，直接判断result的nsp去决定是否跳过该行的计算会让后续开发者难以理解。

File: pkg/sql/plan/function/func_builtin_test.go [hunk 7]
```diff
@@ -759,8 +760,108 @@ func Test_BuiltIn_Math(t *testing.T) {
 			expect: testutil.NewFunctionTestResult(types.T_float64.ToType(), true,
 				nil, nil),
 		}
-		tcc := testutil.NewFunctionTestCase(proc, tc.inputs, tc.expect, builtInLog10)
+		tcc := testutil.NewFunctionTestCase(proc, tc.inputs, tc.expect, TbuiltInLog10)
```
Comment (m-schen) [inline pkg/sql/plan/function/func_builtin_test.go:763]: testutil.NewFunctionTestCase 相关的代码都挪到function目录下的单测文件里吧

File: pkg/sql/plan/function/function.go [hunk 8]
```diff
@@ -488,3 +489,21 @@ func newCheckResultWithCast(overloadId int, castType []types.Type) checkResult {
...
 	}
 }
+
+type FunctionSelectList struct {
+	AnyNull    bool
+	AllNull    bool
+	SelectList []bool
+}
+
+func (selectList *FunctionSelectList) HasAnyNull() bool {
```
Comment (m-schen) [inline pkg/sql/plan/function/function.go:499]: HasAnyNull 和 IsAllNull 需要修改一个名字。
需要体现出 当返回true和false时，是所有的行都应该跳过，还是都不应该跳过。

**Review by m-schen — APPROVED** [review]
加了个not-merge的标签，这两个问题解决后把标签去掉就好。

**Review by heni02 — APPROVED** [review]

**Pull request merged**